nncf/experimental/common/pruning/nodes_grouping.py

```diff
@@ -0,0 +1,465 @@
+    pruning_dimension: int
+
+    @classmethod
+    def from_propagation_block(cls, block: PropagationBlock) -> 'PruningBlock':
+        """
+        Creates an object by taking all necessary information from PropagationBlock.
+        """
+        return cls(block.size, block.offset, block._producer.id, block.pruning_dimension)
+
+    def __str__(self) -> str:
+        return f'S{self.size}_O{self.offset}_PID{self.producer_id}'
+
+    def __hash__(self) -> int:
+        return hash(str(self))
+
+    def __eq__(self, other: 'PruningBlock') -> bool:
+        return str(self) == str(other)
+
+
+@dataclass
+class PruningGroup:
+    """
+    Group of pruning blocks that is obtained after propagation.
+    """
+    dim_blocks: Set[PruningBlock]
+
+    def __eq__(self, other: 'PruningGroup'):
+        return self.dim_blocks == other.dim_blocks
+
+
+def add_group_to_graph(graph: nx.DiGraph,
+                       parent_group: PropagationGroup,
+                       visited_block_ids_map: Dict[int, int]) -> None:
+    """
+    Recursive helper to traverse children of the given PropagationBlock with adding them to the graph.
+
+    :param graph: hierarhy of the propagation groups/blocks.
+    :param root_group: current group for traversing.
+    :param visited_block_ids_map: helper mapping of group addresses to the id in the graph.
+    """
+    global counter
+    parent_graph_id = counter
+    graph.add_node(parent_graph_id, label=str(parent_group))
+    visited_block_ids_map[id(parent_group)] = parent_graph_id
+    if parent_group.has_children():
+        for child_group in parent_group.get_children():
+            child_id = id(child_group)
+            if child_id not in visited_block_ids_map:
+                counter += 1
+                child_graph_id = counter
+                visited_block_ids_map[id(child_group)] = child_graph_id
+                graph.add_edge(parent_graph_id, child_graph_id)
+                add_group_to_graph(graph, child_group, visited_block_ids_map)
+            else:
+                child_graph_id = visited_block_ids_map[child_id]
+                graph.add_edge(parent_graph_id, child_graph_id)
+
+
+counter = 0
+
+
+def build_block_hierarchy(root_groups: List[PropagationGroup]) -> nx.DiGraph:
+    """Creates hierarchy of propagation blocks/groups by traversing children in the given root groups.
+
+    :param roots: list of the root groups
+    :return: networkx graph that represents the hierarchy of propagation blocks/groups.
+    """
+    graph = nx.DiGraph()
+    global counter
+    counter = 0
+    visited = dict()
+    for root_group in root_groups:
+        add_group_to_graph(graph, root_group, visited)
+        counter += 1
+    return graph
+
+
+def get_pruning_groups(graph: NNCFGraph,
+                       pruning_operations_metatypes: PruningOperationsMetatypeRegistry,
+                       prune_operations_types: List[str],
+                       dump_dir: Optional[Path] = None) -> List[PruningGroup]:
```

Comment:
Missing param description for `dump_dir`

Reply:
added